src/test/java/com/amazon/dlic/auth/http/jwt/keybyoidc/SelfRefreshingKeySetTest.java

```diff
@@ -11,108 +11,103 @@
 
-        try {
-            selfRefreshingKeySet.getKey("kid/X");
-            Assert.fail("Expected a BadCredentialsException");
-        } catch (BadCredentialsException e) {
-            Assert.assertEquals(2, selfRefreshingKeySet.getRefreshCount());
-        }
+		JsonWebKey jsonWebKey = selfRefreshingKeySet.getKey("kid/a");
+		Assert.assertEquals(TestJwk.OCT_1_K, jsonWebKey.getKeyProperty("k"));
+	}
 
-    }
+	@Test
+	public void getKey_withNullKidShouldThrowAuthenticatorUnavailableException()
+			throws AuthenticatorUnavailableException, BadCredentialsException {
 
-    @Test(timeout = 10000)
-    public void twoThreadedTest() throws Exception {
-        BlockingMockKeySetProvider provider = new BlockingMockKeySetProvider();
+		Assert.assertThrows(AuthenticatorUnavailableException.class, () -> selfRefreshingKeySet.getKey(null));
 
-        final SelfRefreshingKeySet selfRefreshingKeySet = new SelfRefreshingKeySet(provider);
+	}
 
-        ExecutorService executorService = Executors.newCachedThreadPool();
+	@Test
+	public void getKey_withInvalidDataShouldReturnBadCredentialException()
+			throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        Future<JsonWebKey> f1 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/a"));
+		Assert.assertThrows(BadCredentialsException.class, () -> selfRefreshingKeySet.getKey("kid/X"));
+	}
 
-        provider.waitForCalled();
+	@Test
+	public void getKeyAfterRefresh_withKidShouldReturnKey()
+			throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        Future<JsonWebKey> f2 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/b"));
+		JsonWebKey jsonWebKey = selfRefreshingKeySet.getKeyAfterRefresh("kid/b");
+		Assert.assertEquals(TestJwk.OCT_2_K, jsonWebKey.getKeyProperty("k"));
+	}
 
-        while (selfRefreshingKeySet.getQueuedGetCount() == 0) {
-            Thread.sleep(10);
-        }
+	@Test
+	public void getKeyAfterRefresh_queuedGetCountVariableShouldBeZeroWhenFinishWithAllKeyRefreshes()
+			throws InterruptedException, ExecutionException {
 
-        provider.unblock();
+		int numThreads = 10;
+		ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+		Object lock = new Object();
 
-        Assert.assertEquals(TestJwk.OCT_1_K, f1.get().getProperty("k"));
-        Assert.assertEquals(TestJwk.OCT_2_K, f2.get().getProperty("k"));
+		for (int i = 0; i < numThreads; i++) {
+			executor.submit(() -> {
+				synchronized (lock) {
+					try {
+						selfRefreshingKeySet.getKeyAfterRefresh("kid/a");
+					} catch (AuthenticatorUnavailableException e) {
+					} catch (BadCredentialsException e) {
+					}
+				}
+			});
+		}
 
-        Assert.assertEquals(1, selfRefreshingKeySet.getRefreshCount());
-        Assert.assertEquals(1, selfRefreshingKeySet.getQueuedGetCount());
+		executor.shutdown();
+		executor.awaitTermination(1, TimeUnit.MINUTES);
```

Comment:
Did you find 1 minute from somewhere? I don't know that it is too long or a bad number, but trying to gauge whether we need to keep this so high. In general, I hesitate to add timeouts more then 10 seconds since we end up running tests so often that 1 minute waiting can add up....

Was this from something else or a commonly used value?

Reply:
Sorry, typo here. What I wanted to use was milliseconds.
Now, I realized that it doesn't work with 1 millisecond, so I use 1 second to wait until it finishes.